In [1]:
import os
import mysql.connector
import pandas as pd


In [2]:
mysql_pass=os.environ.get('mysql_password')
#password stored as environment variable for security

db_connection_str = f'mysql+mysqlconnector://root:{mysql_pass}@localhost/bike_share'

def exec_query(query):
    return pd.read_sql(query, con=db_connection_str)

In [3]:
exec_query('SELECT * FROM trips LIMIT 10')

bike_id   checkout_time  duration_minutes  end_station_id  \
0      460 0 days 00:18:49                 4            2501   
1      195 0 days 00:27:51                 2            2563   
2      311 0 days 00:43:41                20            2502   
3      877 0 days 00:43:58                20            2502   
4      273 0 days 01:09:33                16            2566   
5       20 0 days 01:14:52               522            3292   
6      553 0 days 01:15:23                16            3292   
7      432 0 days 01:31:13                40            2711   
8      528 0 days 01:44:07                 8            1008   
9      288 0 days 01:45:21                 3            2495   

                end_station_name  month  start_station_id  \
0                    5th & Bowie      1              1007   
1         Davis at Rainey Street      1              2563   
2     Barton Springs & Riverside      1              2707   
3     Barton Springs & Riverside      1              2707   
4  Pfluger Bridge @ W 2nd Street      1              2569   
5              East 4th & Chicon      1              2495   
6              East 4th & Chicon      1              2495   
7    Barton Springs @ Kinney Ave      1              2536   
8                   Nueces @ 3rd      1              2563   
9                 4th & Congress      1              2498   

                        start_station_name          start_time  \
0                             Lavaca & 6th 2016-01-01 00:18:49   
1                   Davis at Rainey Street 2016-01-01 00:27:51   
2                     Rainey St @ Cummings 2016-01-01 00:43:41   
3                     Rainey St @ Cummings 2016-01-01 00:43:58   
4               East 11th St. & San Marcos 2016-01-01 01:09:33   
5                           4th & Congress 2016-01-01 01:14:52   
6                           4th & Congress 2016-01-01 01:15:23   
7                         Waller & 6th St. 2016-01-01 01:31:13   
8                   Davis at Rainey Street 2016-01-01 01:44:07   
9  Convention Center / 4th St. @ MetroRail 2016-01-01 01:45:21   

  subscriber_type  trip_id  year  
0        Local365  8269930  2016  
1        Local365  8269933  2016  
2         Walk Up  8269945  2016  
3         Walk Up  8269946  2016  
4        Local365  8269949  2016  
5        Local365  8269950  2016  
6        Local365  8269951  2016  
7         Walk Up  8269957  2016  
8        Local365  8269962  2016  
9        Local365  8269963  2016

In [4]:
exec_query('SELECT * FROM stations LIMIT 10')

latitude              location  longitude  \
0   30.2719  (30.27186 -97.73997)   -97.7400   
1   30.2638  (30.26383 -97.72864)   -97.7286   
2   30.2711  (30.27106 -97.74563)   -97.7456   
3   30.2848   (30.2848 -97.72756)   -97.7276   
4   30.2791  (30.27907 -97.73715)   -97.7372   
5   30.2659  (30.26587 -97.76826)   -97.7683   
6   30.2689  (30.26889 -97.74525)   -97.7452   
7   30.2669  (30.26694 -97.74939)   -97.7494   
8   30.2641  (30.26408 -97.74355)   -97.7436   
9   30.2663  (30.26634 -97.74378)   -97.7438   

                                   name  station_id    status  
0               OFFICE/Main/Shop/Repair        1001    active  
1                    6th & Navasota St.        1002    closed  
2                       8th & Guadalupe        1003    closed  
3               Red River & LBJ Library        1004    closed  
4  State Parking Garage @ Brazos & 18th        1005    closed  
5                      Zilker Park West        1006  ACL only  
6                          Lavaca & 6th        1007    active  
7                          Nueces @ 3rd        1008     moved  
8                        2nd & Congress        2494    active  
9                        4th & Congress        2495    active

# Null Values and Data Consistency

The trips and stations tables contain several fields which should be redundant. The end and start station names in the trips table should not be necessary if we have the station ID and the station has an entry in the stations table. The location field in the stations table doesnt seem to provide any information not contained in the latitude and longitude fields. The month, year and checkout_time fields in the trips table also seem to be redundant since the start_time contains the full datetime. Before we normalize the data we will want to investigate the consistency and completeness of the data, so that no useful information is thrown out unnecessarily.

In [5]:
get_num_nulls="""
    SELECT SUM(bike_id IS NULL) AS bike_id,
        SUM(checkout_time IS NULL) AS checkout_time,
        SUM(duration_minutes IS NULL) AS duration_minutes,
        SUM(end_station_id IS NULL) AS end_station_id,
        SUM(end_station_name IS NULL) AS end_station_name,
        SUM(month IS NULL) AS month,
        SUM(start_station_id IS NULL) AS start_station_id,
        SUM(start_station_name IS NULL) AS start_station_name,
        SUM(start_time IS NULL) AS start_time,
        SUM(subscriber_type IS NULL) AS subscriber_type,
        SUM(trip_id IS NULL) AS trip_id,
        SUM(year IS NULL) AS year
    FROM trips"""
exec_query(get_num_nulls)

bike_id  checkout_time  duration_minutes  end_station_id  end_station_name  \
0    723.0            0.0               0.0         19842.0               0.0   

     month  start_station_id  start_station_name  start_time  subscriber_type  \
0  30752.0           19041.0                 0.0         0.0              0.0   

   trip_id     year  
0      0.0  30752.0

We can see that the fields which contain null values are bike_id, end_station_id, start_station_id, month and year. Interestingly both the start and end station names have no null values, so these fields will be important for records with null station IDs. The start_time field also has no null values, so the null years and months may not be an issue. The null bike IDs likely cannot be imputed, but they represent a small fraction of trips and are unlikely to have a major impact on our analysis.

## Date and Time

From the previous query we can see that year and month each have exactly 30752 null values. We may want to know whether these are the same same trips.

In [6]:
missing_year_and_month="""
SELECT COUNT(*) AS missing_year_and_month
FROM trips
WHERE year IS NULL AND month IS NULL"""
exec_query(missing_year_and_month)

missing_year_and_month
0                   30752

There are precisely 30752 records which are missing both the month and year. We are also interested in whether the year and month fields in the remaining records contain data consistent with the start_time

In [7]:
consistent_year="""
SELECT year=YEAR(start_time) AS year_consistent, COUNT(*)
FROM trips
GROUP BY year_consistent"""
exec_query(consistent_year)

year_consistent  COUNT(*)
0              1.0    618479
1              NaN     30752

In [8]:
consistent_month="""
SELECT month=MONTH(start_time) AS month_consistent, COUNT(*)
FROM trips
GROUP BY month_consistent"""
exec_query(consistent_month)

month_consistent  COUNT(*)
0               1.0    618479
1               NaN     30752

As we can see, both the year and month fields are in complete agreement with the start_time field for every trip where year and month are not null. We will want to know if this is also true of the checkout_time.

In [9]:
consistent_time="""
SELECT checkout_time=TIME(start_time) AS time_consistent, COUNT(*)
FROM trips
GROUP BY time_consistent"""
exec_query(consistent_time)

time_consistent  COUNT(*)
0                1    649231

Since the year, month and checkout_time fields do not provide any data different from that in start_time, we will lose nothing by dropping them from the table and relying on start_time.

## Location

In [43]:
latitude="""
    SELECT latitude, CAST(REGEXP_SUBSTR(location,'30.[:digit:]*') AS FLOAT) AS loc_lat
    FROM stations"""
exec_query(latitude)

latitude  loc_lat
0    30.2719  30.2719
1    30.2638  30.2638
2    30.2711  30.2711
3    30.2848  30.2848
4    30.2791  30.2791
..       ...      ...
67   30.2645  30.2645
68   30.2633  30.2633
69   30.2722  30.2722
70   30.2641  30.2641
71   30.2546  30.2546

[72 rows x 2 columns]

In [47]:
longitude="""
    SELECT longitude, CAST(REGEXP_SUBSTR(location,'-97.[:digit:]*') AS FLOAT) AS loc_long
    FROM stations"""
exec_query(longitude)

longitude  loc_long
0    -97.7400  -97.7400
1    -97.7286  -97.7286
2    -97.7456  -97.7456
3    -97.7276  -97.7276
4    -97.7372  -97.7372
..        ...       ...
67   -97.7317  -97.7317
68   -97.7451  -97.7451
69   -97.7525  -97.7525
70   -97.7638  -97.7638
71   -97.7426  -97.7426

[72 rows x 2 columns]

In [44]:
latitude_difference="""
    SELECT latitude - CAST(REGEXP_SUBSTR(location,'30.[:digit:]*') AS FLOAT) AS lat_diff
    FROM stations
    HAVING lat_diff>.00001"""
exec_query(latitude_difference)

Empty DataFrame
Columns: [lat_diff]
Index: []

In [45]:
longitude_difference="""
    SELECT longitude - CAST(REGEXP_SUBSTR(location,'-97.[:digit:]*') AS FLOAT) AS long_diff
    FROM stations
    HAVING long_diff>.00001"""
exec_query(longitude_difference)

Empty DataFrame
Columns: [long_diff]
Index: []

## Station Name and ID

To normalize our database we will have to ensure that station information is contained in the station table and not in the trip table. To do this we will need to ensure a one to one relationship between station ID and station name. We will then be able to use the station name to impute an ID where station ID is null.

In [10]:
ids_per_start_station_name="""
    SELECT start_station_name, COUNT(DISTINCT start_station_id) AS num_ids
    FROM trips
    GROUP BY start_station_name
    ORDER BY num_ids DESC"""
exec_query(ids_per_start_station_name)

start_station_name  num_ids
0                                  11th & San Jacinto        1
1                                  13th & San Antonio        1
2                                    17th & Guadalupe        1
3                                      2nd & Congress        1
4                                          3rd & West        1
..                                                ...      ...
85                         Mobile Station @ Unplugged        0
86                                MapJam at Scoot Inn        0
87                                        Re-branding        0
88  Zilker Park at Barton Springs & William Barton...        0
89                            Red River @ LBJ Library        0

[90 rows x 2 columns]

In [11]:
ids_per_end_station_name="""
    SELECT end_station_name, COUNT(DISTINCT end_station_id) AS num_ids
    FROM trips
    GROUP BY end_station_name
    ORDER BY num_ids DESC"""
exec_query(ids_per_end_station_name)

end_station_name  num_ids
0                                  11th & San Jacinto        1
1                                  13th & San Antonio        1
2                                    17th & Guadalupe        1
3                                      2nd & Congress        1
4                                          3rd & West        1
..                                                ...      ...
87                                     Mobile Station        0
88                                               Shop        0
89                         Mobile Station @ Bike Fest        0
90  Zilker Park at Barton Springs & William Barton...        0
91        Mobile Station @ Boardwalk Opening Ceremony        0

[92 rows x 2 columns]

Thankfully no stations have multiple IDs, but several have none.

In [12]:
start_station_names_per_id="""
SELECT start_station_id, COUNT(DISTINCT start_station_name) AS num_station_names
FROM trips
GROUP BY start_station_id
ORDER BY num_station_names DESC"""
exec_query(start_station_names_per_id)

start_station_id  num_station_names
0                NaN                 19
1             1001.0                  2
2             1002.0                  1
3             1003.0                  1
4             1005.0                  1
..               ...                ...
67            3660.0                  1
68            3684.0                  1
69            3685.0                  1
70            3686.0                  1
71            3687.0                  1

[72 rows x 2 columns]

In [13]:
end_station_names_per_id="""
SELECT end_station_id, COUNT(DISTINCT end_station_name) AS num_station_names
FROM trips
GROUP BY end_station_id
ORDER BY num_station_names DESC"""
exec_query(end_station_names_per_id)

end_station_id  num_station_names
0              NaN                 22
1           1001.0                  1
2           1002.0                  1
3           1003.0                  1
4           1005.0                  1
..             ...                ...
67          3660.0                  1
68          3684.0                  1
69          3685.0                  1
70          3686.0                  1
71          3687.0                  1

[72 rows x 2 columns]

It seems that 1001 is the only ID to have multiple station names.

In [14]:
duplicated_id="""SELECT *
FROM trips
WHERE start_station_id=1001
OR end_station_id=1001"""
exec_query(duplicated_id)

bike_id   checkout_time  duration_minutes  end_station_id  \
0      409 0 days 16:34:57                 4            3455   
1      315 0 days 12:37:33               198            1001   
2      995 0 days 11:12:00                22            2567   
3      621 0 days 17:50:05                47            2542   
4       60 0 days 09:58:24                17            2498   
5      610 0 days 09:31:39               105            2565   
6      281 0 days 13:14:48                 3            2561   
7      957 0 days 15:48:33                12            2566   

                                    end_station_name  month  start_station_id  \
0         Republic Square @ Federal Courthouse Plaza      3              1001   
1                                        Main Office      3              1001   
2                                  Palmer Auditorium      3              1001   
3                                     Plaza Saltillo      3              1001   
4            Convention Center / 4th St. @ MetroRail      3              1001   
5                               Trinity & 6th Street      3              1001   
6  State Capitol Visitors Garage @ San Jacinto & ...      3              1001   
7                      Pfluger Bridge @ W 2nd Street      3              1001   

  start_station_name          start_time subscriber_type   trip_id  year  
0        Main Office 2017-03-01 16:34:57        Local365  13890701  2017  
1        Main Office 2017-03-07 12:37:33        Local365  13948351  2017  
2        Main Office 2017-03-10 11:12:00         Local30  13986174  2017  
3        Main Office 2017-03-10 17:50:05       Weekender  13992957  2017  
4        Main Office 2017-03-15 09:58:24        Local365  14042624  2017  
5             Stolen 2017-03-18 09:31:39        Local365  14088519  2017  
6        Main Office 2017-03-20 13:14:48        Local365  14117889  2017  
7        Main Office 2017-03-27 15:48:33        Local365  14206303  2017

In [15]:
start_stolen="""
SELECT *
FROM trips
WHERE start_station_name='stolen'"""
exec_query(start_stolen)

bike_id   checkout_time  duration_minutes  end_station_id  \
0      610 0 days 09:31:39               105            2565   

       end_station_name  month  start_station_id start_station_name  \
0  Trinity & 6th Street      3              1001             Stolen   

           start_time subscriber_type   trip_id  year  
0 2017-03-18 09:31:39        Local365  14088519  2017

In [16]:
start_station_missing_id="""
    SELECT start_station_name, COUNT(start_station_id) AS num_ids, COUNT(*) AS num_trips
    FROM trips
    GROUP BY start_station_name
    HAVING num_ids < num_trips
    ORDER BY num_trips DESC"""
exec_query(start_station_missing_id)

start_station_name  num_ids  num_trips
0   Zilker Park at Barton Springs & William Barton...        0      11534
1                                   ACC - West & 12th        0       2462
2                    Convention Center/ 3rd & Trinity        0       1292
3                                      Mobile Station        0       1183
4                   East 11th Street at Victory Grill        0       1030
5                             Red River @ LBJ Library        0        584
6                          Mobile Station @ Bike Fest        0        516
7                                         Main Office        7        307
8                               MapJam at Pan Am Park        0         32
9                           MapJam at French Legation        0         27
10                     MapJam at Hops & Grain Brewery        0         19
11                                        Repair Shop        0         13
12                                MapJam at Scoot Inn        0         11
13                                               Shop        0         10
14        Mobile Station @ Boardwalk Opening Ceremony        0          9
15                                        Re-branding        0          7
16                                   Customer Service        0          4
17                         Mobile Station @ Unplugged        0          4
18                                    Marketing Event        0          4

In [17]:
end_station_missing_id="""
    SELECT end_station_name, COUNT(end_station_id) AS num_ids, COUNT(*) AS num_trips
    FROM trips
    GROUP BY end_station_name
    HAVING num_ids < num_trips
    ORDER BY num_trips DESC"""
exec_query(end_station_missing_id)

end_station_name  num_ids  num_trips
0   Zilker Park at Barton Springs & William Barton...        0      13032
1                                   ACC - West & 12th        0       2184
2                    Convention Center/ 3rd & Trinity        0       1460
3                                      Mobile Station        0        905
4                   East 11th Street at Victory Grill        0        761
5                             Red River @ LBJ Library        0        544
6                                         Main Office        1        331
7                                    Customer Service        0        152
8                                         Repair Shop        0        143
9                          Mobile Station @ Bike Fest        0        106
10                                               Shop        0         39
11                              MapJam at Pan Am Park        0         39
12                          MapJam at French Legation        0         29
13                     MapJam at Hops & Grain Brewery        0         26
14                                            Missing        0         25
15                                MapJam at Scoot Inn        0         19
16                                             Stolen        0         15
17                         Mobile Station @ Unplugged        0         13
18                                    Marketing Event        0         12
19        Mobile Station @ Boardwalk Opening Ceremony        0          4
20                                        Re-branding        0          2
21                                          Main Shop        0          2

Of all the station names paired with null station IDs, only the Main Office actually has an ID.

In [50]:
stations_without_id_query="""
    (SELECT DISTINCT start_station_name AS station_missing_id
    FROM trips
    WHERE start_station_id IS NULL)
    UNION DISTINCT
    (SELECT DISTINCT end_station_name
    FROM trips
    WHERE end_station_id IS NULL)"""
stations_without_id=exec_query(stations_without_id_query)
stations_without_id

station_missing_id
0                                         Main Office
1                                         Re-branding
2                                   ACC - West & 12th
3   Zilker Park at Barton Springs & William Barton...
4                             Red River @ LBJ Library
5                                                Shop
6                                         Repair Shop
7                      MapJam at Hops & Grain Brewery
8                   East 11th Street at Victory Grill
9                                    Customer Service
10                                     Mobile Station
11                          MapJam at French Legation
12                              MapJam at Pan Am Park
13                                MapJam at Scoot Inn
14        Mobile Station @ Boardwalk Opening Ceremony
15                   Convention Center/ 3rd & Trinity
16                         Mobile Station @ Unplugged
17                         Mobile Station @ Bike Fest
18                                    Marketing Event
19                                             Stolen
20                                            Missing
21                                          Main Shop

In [52]:
stations_without_id.iloc[3,0]

'Zilker Park at Barton Springs & William Barton Drive'

In [19]:
start_station_sometimes_has_id="""
    SELECT *
    FROM trips t
    INNER JOIN(
        (SELECT DISTINCT start_station_name AS station_missing_id
        FROM trips
        WHERE start_station_id IS NULL)
        UNION DISTINCT
        (SELECT DISTINCT end_station_name
        FROM trips
        WHERE end_station_id IS NULL)) s
    ON t.start_station_name=s.station_missing_id
    WHERE t.start_station_id IS NOT NULL"""
exec_query(end_station_sometimes_has_id)

bike_id   checkout_time  duration_minutes  end_station_id  \
0      409 0 days 16:34:57                 4            3455   
1      315 0 days 12:37:33               198            1001   
2      995 0 days 11:12:00                22            2567   
3      621 0 days 17:50:05                47            2542   
4       60 0 days 09:58:24                17            2498   
5      610 0 days 09:31:39               105            2565   
6      281 0 days 13:14:48                 3            2561   
7      957 0 days 15:48:33                12            2566   

                                    end_station_name  month  start_station_id  \
0         Republic Square @ Federal Courthouse Plaza      3              1001   
1                                        Main Office      3              1001   
2                                  Palmer Auditorium      3              1001   
3                                     Plaza Saltillo      3              1001   
4            Convention Center / 4th St. @ MetroRail      3              1001   
5                               Trinity & 6th Street      3              1001   
6  State Capitol Visitors Garage @ San Jacinto & ...      3              1001   
7                      Pfluger Bridge @ W 2nd Street      3              1001   

  start_station_name          start_time subscriber_type   trip_id  year  \
0        Main Office 2017-03-01 16:34:57        Local365  13890701  2017   
1        Main Office 2017-03-07 12:37:33        Local365  13948351  2017   
2        Main Office 2017-03-10 11:12:00         Local30  13986174  2017   
3        Main Office 2017-03-10 17:50:05       Weekender  13992957  2017   
4        Main Office 2017-03-15 09:58:24        Local365  14042624  2017   
5             Stolen 2017-03-18 09:31:39        Local365  14088519  2017   
6        Main Office 2017-03-20 13:14:48        Local365  14117889  2017   
7        Main Office 2017-03-27 15:48:33        Local365  14206303  2017   

  station_missing_id  
0        Main Office  
1        Main Office  
2        Main Office  
3        Main Office  
4        Main Office  
5             Stolen  
6        Main Office  
7        Main Office

In [20]:
end_station_sometimes_has_id="""
    SELECT *
    FROM trips t
    INNER JOIN(
        (SELECT DISTINCT start_station_name AS station_missing_id
        FROM trips
        WHERE start_station_id IS NULL)
        UNION DISTINCT
        (SELECT DISTINCT end_station_name
        FROM trips
        WHERE end_station_id IS NULL)) s
    ON t.end_station_name=s.station_missing_id
    WHERE t.end_station_id IS NOT NULL"""
exec_query(start_station_sometimes_has_id)

bike_id   checkout_time  duration_minutes  end_station_id end_station_name  \
0      315 0 days 12:37:33               198            1001      Main Office   

   month  start_station_id start_station_name          start_time  \
0      3              1001        Main Office 2017-03-07 12:37:33   

  subscriber_type   trip_id  year station_missing_id  
0        Local365  13948351  2017        Main Office

For 